File: src/main/scala/li/cil/oc/util/MovingAverage.scala
package li.cil.oc.util

class MovingAverage(val size: Int) {
  private val data = Array.fill(size)(0)
  private var head = 0
  private var cachedAverage = 0
  private var dirty = true

  def apply() = {
    if (dirty) {
      cachedAverage = data.sum / size
      dirty = false
    }
    cachedAverage
  }

  def +=(value: Int) = {
    data(head) = value
    head = (head + 1) % size
    dirty = true
    this
  }
}

File: src/main/scala/li/cil/oc/util/MovingAverage.java
package li.cil.oc.util;

public final class MovingAverage {
    private final int[] data;
    private int head;
    private int cachedAverage;
    private boolean dirty = true;

    // ----------------------------------------------------------------------- //

    public MovingAverage(final int size) {
        data = new int[size];
    }

    public int get() {
        if (dirty) {
            int sum = 0;
            for (final int value : data) {
                sum += value;
            }
            cachedAverage = sum / data.length;
            dirty = false;
        }
        return cachedAverage;
    }

    public void put(final int value) {
        data[head] = value;
        head = (head + 1) % data.length;
        dirty = true;
    }
}
